Geometry/HGCalCommonData/src/HGCalGeomParameters.cc

```diff
@@ -375,72 +378,227 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
-	      << ", " << php.trformTranY_[k] << ", " << php.trformTranZ_[k] 
-	      << " Rotation ("  << php.trformRotXX_[k] << ", "
-	      << php.trformRotYX_[k] << ", " << php.trformRotZX_[k] << ", "
-	      << php.trformRotXY_[k] << ", " << php.trformRotYY_[k] << ", "
-	      << php.trformRotZY_[k] << ", " << php.trformRotXZ_[k] << ", "
-	      << php.trformRotYZ_[k] << ", " << php.trformRotZZ_[k] << ")\n";
+    edm::LogVerbatim("HGCalGeom") << "Matrix[" << k << "] (" << std::hex 
+				  << php.trformIndex_[k]
+				  << std::dec << ") Trnaslation (" 
+				  << php.trformTranX_[k] << ", " 
+				  << php.trformTranY_[k] << ", " 
+				  << php.trformTranZ_[k] << " Rotation ("
+				  << php.trformRotXX_[k] << ", "
+				  << php.trformRotYX_[k] << ", " 
+				  << php.trformRotZX_[k] << ", "
+				  << php.trformRotXY_[k] << ", " 
+				  << php.trformRotYY_[k] << ", "
+				  << php.trformRotZY_[k] << ", " 
+				  << php.trformRotXZ_[k] << ", "
+				  << php.trformRotYZ_[k] << ", " 
+				  << php.trformRotZZ_[k] << ")";
   }
-  std::cout << "Dump copiesInLayers for " << php.copiesInLayers_.size()
-	    << " layers\n";
+  edm::LogVerbatim("HGCalGeom") << "Dump copiesInLayers for " 
+				<< php.copiesInLayers_.size()
+				<< " layers";
   for (unsigned int k=0; k<php.copiesInLayers_.size(); ++k) {
     const auto& theModules = php.copiesInLayers_[k];
-    std::cout << "Layer " << k << ":" << theModules.size() << std::endl;
+    edm::LogVerbatim("HGCalGeom") << "Layer " << k << ":" <<theModules.size();
     int k2(0);
     for (std::unordered_map<int, int>::const_iterator itr=theModules.begin();
-	 itr != theModules.end(); ++itr) {
-      std::cout << " " << itr->first << ":" << itr->second;
-      ++k2;
-      if (k2 > 9) { std::cout << std::endl; k2 = 0; }
+	 itr != theModules.end(); ++itr, ++k2) {
+      edm::LogVerbatim("HGCalGeom") << "[" << k2 << "] " << itr->first << ":" 
+				    << itr->second;
     }
-    if (k2 > 0) std::cout << std::endl;
+  }
+#endif
+}
+
+void HGCalGeomParameters::loadGeometryHexagon8(const DDFilteredView& _fv, 
+					       HGCalParameters& php) {
+ 
+  DDFilteredView fv = _fv;
+  bool dodet(true);
+  std::map<int,HGCalGeomParameters::layerParameters> layers;
+  std::vector<HGCalParameters::hgtrform> trforms;
+  std::vector<bool>                      trformUse;
+  
+  while (dodet) {
+    const DDSolid & sol  = fv.logicalPart().solid();
+    std::string     name = sol.name();
+    // Layers first
+    std::vector<int> copy = fv.copyNumbers();
+    int nsiz = (int)(copy.size());
+    int lay  = (nsiz > 0) ? copy[nsiz-1] : 0;
+    int zp   = (nsiz > 3) ? copy[3] : -1;
+    if (zp != 1) zp = -1;
+    if (lay == 0) {
+      edm::LogError("HGCalGeom") << "Funny layer # " << lay << " zp "
+				 << zp << " in " << nsiz << " components";
+      throw cms::Exception("DDException") << "Funny layer # " << lay;
+    } else {
+      if (std::find(php.layer_.begin(),php.layer_.end(),lay) == 
```

Comment:
how often is this `find()` going to be called, and how large is `php.layer_`? Consider using a more efficient search, if it's possible to sort `php.layer_`.

Reply:
It does not come in event loop - only during initialization

Reply:
It does not come in event loop - only during initialization